radix-engine/src/model/resource.rs

```diff
@@ -0,0 +1,282 @@
+    UnsupportedOperation,
+}
+
+/// Represents the contained resource.
+#[derive(Debug, Clone, TypeId, Encode, Decode)]
+pub enum Resource {
+    // TODO: update resource state based on proofs.
+    Fungible {
+        /// The locked amounts and the corresponding times of being locked.
+        locked_amounts: BTreeMap<Decimal, usize>,
+        /// The liquid amount.
+        liquid_amount: Decimal,
+    },
+    NonFungible {
+        /// The locked non-fungible ids and the corresponding times of being locked.
+        locked_ids: HashMap<NonFungibleId, usize>,
+        /// The liquid non-fungible ids.
+        liquid_ids: BTreeSet<NonFungibleId>,
+    },
+}
+
+#[derive(Debug, Clone, TypeId, Encode, Decode)]
+pub enum ResourceAmount {
+    /// Fungible amount
+    Fungible { amount: Decimal },
+    /// Non-fungible amount
+    NonFungible { ids: BTreeSet<NonFungibleId> },
+}
+
+/// A transient resource container.
+#[derive(Debug, TypeId, Encode, Decode)]
+pub struct ResourceContainer {
+    /// The resource definition id
+    resource_def_id: ResourceDefId,
+    /// The resource type
+    resource_type: ResourceType,
+    /// The contained resource
+    resource: Resource,
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum ResourceContainerId {
+    /// For named bucket
+    Bucket(BucketId),
+    /// For vault
+    Vault(VaultId),
+
+    /// For the specific resource on the n-th worktop
+    Worktop {
+        depth: u32,
+        resource_def_id: ResourceDefId,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub struct Proof {
+    /// The resource definition id
+    resource_def_id: ResourceDefId,
+    /// The resource type
+    resource_type: ResourceType,
+    /// Restricted proof can't be moved down along the call stack (growing down).
+    restricted: bool,
+    /// The total amount locked
+    total_amount: ResourceAmount,
+    /// The sub-amounts (to be extended)
+    #[allow(dead_code)]
+    amounts: HashMap<ResourceContainerId, (Rc<ResourceContainer>, ResourceAmount)>,
+}
+
+impl Resource {
+    pub fn fungible(amount: Decimal) -> Self {
+        Resource::Fungible {
+            locked_amounts: BTreeMap::new(),
+            liquid_amount: amount,
+        }
+    }
+
+    pub fn non_fungible(ids: BTreeSet<NonFungibleId>) -> Self {
+        Resource::NonFungible {
+            locked_ids: HashMap::new(),
+            liquid_ids: ids.clone(),
+        }
+    }
+}
+
+impl ResourceAmount {
+    pub fn quantity(&self) -> Decimal {
+        match self {
+            ResourceAmount::Fungible { amount } => *amount,
+            ResourceAmount::NonFungible { ids } => ids.len().into(),
+        }
+    }
+
+    pub fn non_fungible_ids(&self) -> Result<BTreeSet<NonFungibleId>, ResourceError> {
+        match self {
+            ResourceAmount::Fungible { .. } => Err(ResourceError::UnsupportedOperation),
+            ResourceAmount::NonFungible { ids } => Ok(ids.clone()),
+        }
+    }
+}
+
+impl ResourceContainer {
```

Comment:
Not sure if this is the right abstraction here as I think the way vaults and buckets implement `take` would be quite different I think

Reply:
Perhaps this would be better as a trait?

Reply:
The main difference I see is authorization and liquid amount representation. The abstraction here is to ensure the same accounting logic used by all containers. Make it a trait sounds a bit off to me.

A better solution is to have `ResourceContainer<T>` where `T` is any struct with trait for offering liquidity.